File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/vertx/VertxRepositoryGenerator.java
package ru.tinkoff.kora.database.annotation.processor.vertx;

import com.palantir.javapoet.*;
import jakarta.annotation.Nullable;
import ru.tinkoff.kora.annotation.processor.common.*;
import ru.tinkoff.kora.common.Tag;
import ru.tinkoff.kora.database.annotation.processor.DbUtils;
import ru.tinkoff.kora.database.annotation.processor.QueryWithParameters;
import ru.tinkoff.kora.database.annotation.processor.RepositoryGenerator;
import ru.tinkoff.kora.database.annotation.processor.model.QueryParameter;
import ru.tinkoff.kora.database.annotation.processor.model.QueryParameterParser;

import javax.annotation.processing.Filer;
import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.ExecutableType;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.*;
import java.util.concurrent.CompletionStage;

public final class VertxRepositoryGenerator implements RepositoryGenerator {
    private final TypeMirror repositoryInterface;
    private final Types types;
    private final Elements elements;
    private final Filer filer;
    private final DeclaredType completionStageType;

    public VertxRepositoryGenerator(ProcessingEnvironment processingEnv) {
        var repository = processingEnv.getElementUtils().getTypeElement(VertxTypes.REPOSITORY.canonicalName());
        if (repository == null) {
            this.repositoryInterface = null;
        } else {
            this.repositoryInterface = repository.asType();
        }
        this.types = processingEnv.getTypeUtils();
        this.elements = processingEnv.getElementUtils();
        this.filer = processingEnv.getFiler();
        this.completionStageType = this.types.getDeclaredType(
            this.elements.getTypeElement(CompletionStage.class.getCanonicalName()),
            this.types.getWildcardType(null, null)
        );
    }

    @Override
    @Nullable
    public TypeMirror repositoryInterface() {
        return this.repositoryInterface;
    }

    @Override
    public TypeSpec generate(TypeElement repositoryElement, TypeSpec.Builder type, MethodSpec.Builder constructor) {
        var repositoryType = (DeclaredType) repositoryElement.asType();
        var queryMethods = DbUtils.findQueryMethods(this.types, this.elements, repositoryElement);
        this.enrichWithExecutor(repositoryElement, type, constructor, queryMethods);
        var resultMappers = new FieldFactory(this.types, elements, type, constructor, "_result_mapper_");
        var parameterMappers = new FieldFactory(this.types, elements, type, constructor, "_parameter_mapper_");

        int methodCounter = 1;
        for (var method : queryMethods) {
            var methodType = (ExecutableType) this.types.asMemberOf(repositoryType, method);
            var parameters = QueryParameterParser.parse(this.types, List.of(VertxTypes.CONNECTION, VertxTypes.SQL_CLIENT), VertxTypes.PARAMETER_COLUMN_MAPPER, method, methodType);
            var queryAnnotation = AnnotationUtils.findAnnotation(method, DbUtils.QUERY_ANNOTATION);
            var queryString = AnnotationUtils.<String>parseAnnotationValueWithoutDefault(queryAnnotation, "value");
            var query = QueryWithParameters.parse(filer, types, queryString, parameters, repositoryType, method);
            var resultMapper = this.parseResultMapper(method, parameters, methodType)
                .map(rm -> DbUtils.addMapper(resultMappers, rm))
                .orElse(null);
            DbUtils.addMappers(parameterMappers, DbUtils.parseParameterMappers(
                parameters,
                query,
                tn -> VertxNativeTypes.find(tn) != null,
                VertxTypes.PARAMETER_COLUMN_MAPPER
            ));
            var methodSpec = this.generate(type, methodCounter, method, methodType, query, parameters, resultMapper, parameterMappers);
            type.addMethod(methodSpec);
            methodCounter++;
        }
        return type.addMethod(constructor.build()).build();
    }

    private MethodSpec generate(TypeSpec.Builder type, int methodNumber, ExecutableElement method, ExecutableType methodType, QueryWithParameters query, List<QueryParameter> parameters, @Nullable String resultMapperName, FieldFactory parameterMappers) {
        var sql = query.rawQuery();
        {
            var params = new ArrayList<Map.Entry<QueryWithParameters.QueryParameter, Integer>>(query.parameters().size());
            for (int i = 0; i < query.parameters().size(); i++) {
                var parameter = query.parameters().get(i);
                params.add(Map.entry(parameter, i));
            }
            for (var parameter : params.stream().sorted(Comparator.<Map.Entry<QueryWithParameters.QueryParameter, Integer>, Integer>comparing(p -> p.getKey().sqlParameterName().length()).reversed()).toList()) {
                sql = sql.replace(":" + parameter.getKey().sqlParameterName(), "$" + (parameter.getValue() + 1));
            }
        }
        var b = DbUtils.queryMethodBuilder(method, methodType);

        var queryContextFieldName = "QUERY_CONTEXT_" + methodNumber;
        type.addField(
            FieldSpec.builder(DbUtils.QUERY_CONTEXT, queryContextFieldName, Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
                .initializer("""
                    new $T(
                          $S,
                          $S,
                          $S
                        )""", DbUtils.QUERY_CONTEXT, query.rawQuery(), sql, DbUtils.operationName(method))
                .build());
        b.addStatement("var _query = $L", queryContextFieldName);

        var batchParam = parameters.stream().filter(QueryParameter.BatchParameter.class::isInstance).findFirst().orElse(null);
        var connectionParam = parameters.stream().filter(QueryParameter.ConnectionParameter.class::isInstance).findFirst().orElse(null);
        var returnType = methodType.getReturnType();
        var isFlux = CommonUtils.isFlux(returnType);
        var isMono = CommonUtils.isMono(returnType);
        var isCompletionStage = this.isCompletionStage(returnType);
        var isVoid = isVoid(returnType);

        ParametersToTupleBuilder.generate(b, query, method, parameters, batchParam, parameterMappers);
        CodeBlock resultMapper;
        if (isVoid) {
            resultMapper = CodeBlock.of("_rs -> null");
        } else if ((isMono || isCompletionStage) && ((DeclaredType) returnType).getTypeArguments().get(0).toString().equals(DbUtils.UPDATE_COUNT.canonicalName())) {
            resultMapper = CodeBlock.of("$T::extractUpdateCount", VertxTypes.ROW_SET_MAPPER);
        } else if (returnType.toString().equals(DbUtils.UPDATE_COUNT.canonicalName())) {
            resultMapper = CodeBlock.of("$T::extractUpdateCount", VertxTypes.ROW_SET_MAPPER);
        } else {
            Objects.requireNonNull(resultMapperName, () -> "Illegal State occurred when expected to get result mapper, but got null in " + method.getEnclosingElement().getSimpleName() + "#" + method.getSimpleName());
            resultMapper = CodeBlock.of("$N", resultMapperName);
        }
        if (returnType.getKind() != TypeKind.VOID) {
            b.addCode("return ");
        }
        if (batchParam != null) {
            if (isCompletionStage || !isMono) {
                if (connectionParam == null) {
                    b.addCode("$T.batchCompletionStage(this._connectionFactory, _query, _batchParams)\n", VertxTypes.REPOSITORY_HELPER);
                } else {
                    b.addCode("$T.batchCompletionStage($N, this._connectionFactory.telemetry(), _query, _batchParams)\n", VertxTypes.REPOSITORY_HELPER, connectionParam.name());
                }
                if (isVoid) {
                    b.addCode("  .thenApply(v -> (Void) null)\n");
                }
            } else {
                if (connectionParam == null) {
                    b.addCode("$T.Reactor.batchMono(this._connectionFactory, _query, _batchParams)\n", VertxTypes.REPOSITORY_HELPER);
                } else {
                    b.addCode("$T.Reactor.batchMono($N, this._connectionFactory.telemetry(), _query, _batchParams)\n", VertxTypes.REPOSITORY_HELPER, connectionParam.name());
                }
                if (isVoid) {
                    b.addCode("  .then()\n");
                }
            }
        } else if (isFlux) {
            if (connectionParam == null) {
                b.addCode("$T.Reactor.flux(this._connectionFactory, _query, _tuple, $L)\n", VertxTypes.REPOSITORY_HELPER, resultMapperName);
            } else {
                b.addCode("$T.Reactor.flux($N, this._connectionFactory.telemetry(), _query, _tuple, $L)\n", VertxTypes.REPOSITORY_HELPER, connectionParam.name(), resultMapperName);
            }
        } else if (isMono) {
            if (connectionParam == null) {
                b.addCode("$T.Reactor.mono(this._connectionFactory, _query, _tuple, $L)\n", VertxTypes.REPOSITORY_HELPER, resultMapper);
            } else {
                b.addCode("$T.Reactor.mono($N, this._connectionFactory.telemetry(), _query, _tuple, $L)\n", VertxTypes.REPOSITORY_HELPER, connectionParam.name(), resultMapper);
            }
        } else {
            if (connectionParam == null) {
                b.addCode("$T.completionStage(this._connectionFactory, _query, _tuple, $L)\n", VertxTypes.REPOSITORY_HELPER, resultMapper);
            } else {
                b.addCode("$T.completionStage($N, this._connectionFactory.telemetry(), _query, _tuple, $L)\n", VertxTypes.REPOSITORY_HELPER, connectionParam.name(), resultMapper);
            }
        }
        if (isFlux) {
            b.addCode(";\n");
        } else if (isMono) {
            b.addCode(";\n");
        } else if (isCompletionStage) {
            b.addCode(";\n");
        } else {
            b.addCode("  .join();\n");
        }
        return b.build();
    }

    private Optional<DbUtils.Mapper> parseResultMapper(ExecutableElement method, List<QueryParameter> parameters, ExecutableType methodType) {
        var returnType = methodType.getReturnType();
        if (isVoid(returnType)) {
            return Optional.empty();
        }

        final boolean isFlux = CommonUtils.isFlux(returnType);
        final boolean isMono = CommonUtils.isMono(returnType);
        final boolean isCompletionStage = isCompletionStage(returnType);
        for (var parameter : parameters) {
            if (parameter instanceof QueryParameter.BatchParameter) {
                final TypeMirror realReturnType = (isMono || isFlux || isCompletionStage)
                    ? Visitors.visitDeclaredType(returnType, dt -> dt.getTypeArguments().get(0))
                    : method.getReturnType();

                if (CommonUtils.isVoid(realReturnType)) {
                    return Optional.empty();
                } else if (DbUtils.UPDATE_COUNT.canonicalName().equals(realReturnType.toString())) {
                    return Optional.empty();
                } else {
                    throw new ProcessingErrorException("@Batch method can't return arbitrary values, it can only return: void/UpdateCount", method);
                }
            }
        }

        var mappings = CommonUtils.parseMapping(method);
        var rowSetMapper = mappings.getMapping(VertxTypes.ROW_SET_MAPPER);
        var rowMapper = mappings.getMapping(VertxTypes.ROW_MAPPER);
        if (isFlux) {
            var fluxParam = Visitors.visitDeclaredType(returnType, dt -> dt.getTypeArguments().get(0));
            var mapperType = ParameterizedTypeName.get(VertxTypes.ROW_MAPPER, TypeName.get(fluxParam));
            if (rowMapper != null) {
                return Optional.of(new DbUtils.Mapper(rowMapper.mapperClass(), mapperType, rowMapper.mapperTags()));
            }
            return Optional.of(new DbUtils.Mapper(mapperType, Set.of()));
        }
        if (isMono || isCompletionStage) {
            var monoParam = Visitors.visitDeclaredType(returnType, dt -> dt.getTypeArguments().get(0));
            var mapperType = ParameterizedTypeName.get(VertxTypes.ROW_SET_MAPPER, TypeName.get(monoParam));
            if (monoParam.toString().equals(DbUtils.UPDATE_COUNT.canonicalName())) {
                return Optional.empty();
            }
            if (rowSetMapper != null) {
                return Optional.of(new DbUtils.Mapper(rowSetMapper.mapperClass(), mapperType, rowSetMapper.mapperTags()));
            }
            if (rowMapper != null) {
                if (CommonUtils.isList(monoParam)) {
                    return Optional.of(new DbUtils.Mapper(rowMapper.mapperClass(), mapperType, rowMapper.mapperTags(), c -> CodeBlock.of("$T.listRowSetMapper($L)", VertxTypes.ROW_SET_MAPPER, c)));
                } else if (CommonUtils.isOptional(monoParam)) {
                    return Optional.of(new DbUtils.Mapper(rowMapper.mapperClass(), mapperType, rowMapper.mapperTags(), c -> CodeBlock.of("$T.optionalRowSetMapper($L)", VertxTypes.ROW_SET_MAPPER, c)));
                } else {
                    return Optional.of(new DbUtils.Mapper(rowMapper.mapperClass(), mapperType, rowMapper.mapperTags(), c -> CodeBlock.of("$T.singleRowSetMapper($L)", VertxTypes.ROW_SET_MAPPER, c)));
                }
            }
            return Optional.of(new DbUtils.Mapper(mapperType, Set.of()));
        }
        var mapperType = ParameterizedTypeName.get(VertxTypes.ROW_SET_MAPPER, TypeName.get(returnType).box());
        if (rowSetMapper != null) {
            return Optional.of(new DbUtils.Mapper(rowSetMapper.mapperClass(), mapperType, rowSetMapper.mapperTags()));
        }
        if (rowMapper != null) {
            if (CommonUtils.isList(returnType)) {
                return Optional.of(new DbUtils.Mapper(rowMapper.mapperClass(), mapperType, rowMapper.mapperTags(), c -> CodeBlock.of("$T.listRowSetMapper($L)", VertxTypes.ROW_SET_MAPPER, c)));
            } else if (CommonUtils.isOptional(returnType)) {
                return Optional.of(new DbUtils.Mapper(rowMapper.mapperClass(), mapperType, rowMapper.mapperTags(), c -> CodeBlock.of("$T.optionalRowSetMapper($L)", VertxTypes.ROW_SET_MAPPER, c)));
            } else {
                return Optional.of(new DbUtils.Mapper(rowMapper.mapperClass(), mapperType, rowMapper.mapperTags(), c -> CodeBlock.of("$T.singleRowSetMapper($L)", VertxTypes.ROW_SET_MAPPER, c)));
            }
        }
        if (returnType.toString().equals(DbUtils.UPDATE_COUNT.canonicalName())) {
            return Optional.empty();
        }
        return Optional.of(new DbUtils.Mapper(mapperType, Set.of()));
    }

    public void enrichWithExecutor(TypeElement repositoryElement, TypeSpec.Builder builder, MethodSpec.Builder constructorBuilder, List<ExecutableElement> queryMethods) {
        builder.addField(VertxTypes.CONNECTION_FACTORY, "_connectionFactory", Modifier.PRIVATE, Modifier.FINAL);
        builder.addSuperinterface(VertxTypes.REPOSITORY);
        builder.addMethod(MethodSpec.methodBuilder("getVertxConnectionFactory")
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addAnnotation(Override.class)
            .returns(VertxTypes.CONNECTION_FACTORY)
            .addCode("return this._connectionFactory;")
            .build());

        var executorTag = DbUtils.getTag(repositoryElement);
        if (executorTag != null) {
            constructorBuilder.addParameter(ParameterSpec.builder(VertxTypes.CONNECTION_FACTORY, "_connectionFactory").addAnnotation(AnnotationSpec.builder(Tag.class).addMember("value", executorTag).build()).build());
        } else {
            constructorBuilder.addParameter(VertxTypes.CONNECTION_FACTORY, "_connectionFactory");
        }
        constructorBuilder.addStatement("this._connectionFactory = _connectionFactory");
    }

    private boolean isCompletionStage(TypeMirror returnType) {
        return this.types.isAssignable(returnType, this.completionStageType);
    }

    private boolean isVoid(TypeMirror tm) {
        if (isCompletionStage(tm) || CommonUtils.isMono(tm) || CommonUtils.isFlux(tm)) {
            tm = Visitors.visitDeclaredType(tm, dt -> dt.getTypeArguments().get(0));
        }
        return CommonUtils.isVoid(tm);
    }
}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/vertx/VertxRepositoryGenerator.kt
package ru.tinkoff.kora.database.symbol.processor.vertx

import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSFunction
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.toTypeName
import ru.tinkoff.kora.database.symbol.processor.DbUtils
import ru.tinkoff.kora.database.symbol.processor.DbUtils.addMapper
import ru.tinkoff.kora.database.symbol.processor.DbUtils.asFlow
import ru.tinkoff.kora.database.symbol.processor.DbUtils.findQueryMethods
import ru.tinkoff.kora.database.symbol.processor.DbUtils.operationName
import ru.tinkoff.kora.database.symbol.processor.DbUtils.parseExecutorTag
import ru.tinkoff.kora.database.symbol.processor.DbUtils.queryMethodBuilder
import ru.tinkoff.kora.database.symbol.processor.DbUtils.resultMapperName
import ru.tinkoff.kora.database.symbol.processor.DbUtils.updateCount
import ru.tinkoff.kora.database.symbol.processor.Mapper
import ru.tinkoff.kora.database.symbol.processor.QueryWithParameters
import ru.tinkoff.kora.database.symbol.processor.RepositoryGenerator
import ru.tinkoff.kora.database.symbol.processor.model.QueryParameter
import ru.tinkoff.kora.database.symbol.processor.model.QueryParameterParser
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findValue
import ru.tinkoff.kora.ksp.common.CommonClassNames
import ru.tinkoff.kora.ksp.common.CommonClassNames.isFlow
import ru.tinkoff.kora.ksp.common.CommonClassNames.isList
import ru.tinkoff.kora.ksp.common.FieldFactory
import ru.tinkoff.kora.ksp.common.FunctionUtils.isFlow
import ru.tinkoff.kora.ksp.common.FunctionUtils.isSuspend
import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException
import ru.tinkoff.kora.ksp.common.parseMappingData
import java.util.concurrent.CompletableFuture

class VertxRepositoryGenerator(private val resolver: Resolver, private val kspLogger: KSPLogger) : RepositoryGenerator {
    private val repositoryInterface = resolver.getClassDeclarationByName(resolver.getKSNameFromString(VertxTypes.repository.canonicalName))?.asStarProjectedType()
    override fun repositoryInterface() = repositoryInterface

    override fun generate(repositoryType: KSClassDeclaration, typeBuilder: TypeSpec.Builder, constructorBuilder: FunSpec.Builder): TypeSpec {
        this.enrichWithExecutor(repositoryType, typeBuilder, constructorBuilder)
        val repositoryResolvedType = repositoryType.asStarProjectedType()
        val resultMappers = FieldFactory(typeBuilder, constructorBuilder, "_result_mapper_")
        val parameterMappers = FieldFactory(typeBuilder, constructorBuilder, "_parameter_mapper_")

        var methodCounter = 1
        for (method in repositoryType.findQueryMethods()) {
            val methodType = method.asMemberOf(repositoryResolvedType)
            val parameters = QueryParameterParser.parse(listOf(VertxTypes.sqlConnection, VertxTypes.sqlClient), VertxTypes.parameterColumnMapper, method, methodType)
            val queryAnnotation = method.findAnnotation(DbUtils.queryAnnotation)!!
            val queryString = queryAnnotation.findValue<String>("value")!!
            val query = QueryWithParameters.parse(queryString, parameters, method)
            val resultMapperName = this.parseResultMapper(method, parameters, methodType)?.let { resultMappers.addMapper(it) }
            DbUtils.parseParameterMappers(method, parameters, query, VertxTypes.parameterColumnMapper) { VertxNativeTypes.findNativeType(it.toTypeName()) != null }
                .forEach { parameterMappers.addMapper(it) }
            val methodSpec = this.generate(typeBuilder, methodCounter, method, methodType, query, parameters, resultMapperName, parameterMappers)
            typeBuilder.addFunction(methodSpec)
            methodCounter++
        }

        return typeBuilder.primaryConstructor(constructorBuilder.build()).build()
    }

    private data class QueryReplace(val index: Int, val paramIndex: Int, val name: String)

    private fun generate(
        typeBuilder: TypeSpec.Builder,
        methodNumber: Int,
        funDeclaration: KSFunctionDeclaration,
        function: KSFunction,
        query: QueryWithParameters,
        parameters: List<QueryParameter>,
        resultMapperName: String?,
        parameterMappers: FieldFactory
    ): FunSpec {
        var sql = query.rawQuery

        val replaceParams = mutableListOf<QueryReplace>()
        for (i in query.parameters.indices) {
            val parameter = query.parameters[i]
            for (queryIndex in parameter.queryIndexes) {
                replaceParams.add(QueryReplace(queryIndex, i + 1, parameter.sqlParameterName))
            }
        }
        replaceParams.sortBy { it.index }
        var sqlIndexDiff = 0
        for (parameter in replaceParams) {
            val queryIndexAdjusted: Int = parameter.index - sqlIndexDiff
            sql = sql.substring(0, queryIndexAdjusted) + "$" + parameter.paramIndex + sql.substring(queryIndexAdjusted + parameter.name.length + 1)
            sqlIndexDiff += (parameter.name.length - parameter.paramIndex.toString().length)
        }

        val b = funDeclaration.queryMethodBuilder(resolver)

        val queryContextFieldName = "_queryContext_$methodNumber"
        typeBuilder.addProperty(
            PropertySpec.builder(queryContextFieldName, DbUtils.queryContext, KModifier.PRIVATE)
                .initializer(
                    """
                    %T(
                      %S,
                      %S,
                      %S
                    )
                    """.trimIndent(), DbUtils.queryContext, query.rawQuery, sql, funDeclaration.operationName()
                )
                .build()
        )
        b.addStatement("val _query = %L", queryContextFieldName)

        val batchParam = parameters.firstOrNull { it is QueryParameter.BatchParameter }
        val isSuspend = funDeclaration.isSuspend()
        val isFlow = funDeclaration.isFlow()
        ParametersToTupleBuilder.generate(b, query, parameters, batchParam, parameterMappers)
        val connectionParameter = parameters.asSequence().filterIsInstance<QueryParameter.ConnectionParameter>().firstOrNull()?.variable?.name?.asString()

        if (batchParam != null) {
            b.addCode("var _future = ")
            if (connectionParameter == null) {
                b.addCode("%T.batchCompletionStage(this._vertxConnectionFactory, _query, _batchParams)\n", VertxTypes.repositoryHelper)
            } else {
                b.addCode("%T.batchCompletionStage(%N, this._vertxConnectionFactory.telemetry(), _query, _batchParams)\n", VertxTypes.repositoryHelper, connectionParameter)
            }
            if (function.returnType == resolver.builtIns.unitType) {
                b.addCode("  .thenApply {}\n")
            }
        } else if (isFlow) {
            b.addCode("return ")
            if (connectionParameter == null) {
                b.addCode("%T.Flux.flux(this._vertxConnectionFactory, _query, _tuple, %N).%M()\n", VertxTypes.repositoryHelper, resultMapperName, asFlow)
            } else {
                b.addCode("%T.Flux.flux(%N, this._vertxConnectionFactory.telemetry(), _query, _tuple, %N).%M()\n", VertxTypes.repositoryHelper, connectionParameter, resultMapperName, asFlow)
            }
            return b.build()
        } else {
            b.addCode("var _future: %T = ", CompletableFuture::class.asClassName().parameterizedBy(function.returnType!!.toTypeName()))
            if (connectionParameter == null) {
                b.addCode("%T.completionStage(this._vertxConnectionFactory, _query, _tuple", VertxTypes.repositoryHelper)
            } else {
                b.addCode("%T.completionStage(%N, this._vertxConnectionFactory.telemetry(), _query, _tuple", VertxTypes.repositoryHelper, connectionParameter)
            }

            if (function.returnType == resolver.builtIns.unitType) {
                b.addCode(") {}\n")
            } else if (function.returnType?.toTypeName() == updateCount) {
                b.addCode(") { %T.extractUpdateCount(it) }\n", VertxTypes.rowSetMapper)
            } else {
                b.addCode(", %N)\n", resultMapperName)
            }

        }
        if (isSuspend) {
            b.addCode("return _future.%M()", CommonClassNames.await)
        } else if (!isFlow) {
            b.addCode("return _future.toCompletableFuture().join()")
            if (function.returnType!!.isMarkedNullable) {
                b.addCode("!!")
            }
        }
        return b.build()
    }

    private fun parseResultMapper(method: KSFunctionDeclaration, parameters: List<QueryParameter>, methodType: KSFunction): Mapper? {
        val returnType = methodType.returnType!!
        val returnTypeName = returnType.toTypeName().copy(false)
        if (returnTypeName == UNIT) {
            return null
        }
        if (returnTypeName == updateCount) {
            return null
        }

        for (parameter in parameters) {
            if (parameter is QueryParameter.BatchParameter) {
                throw ProcessingErrorException("@Batch method can't return arbitrary values, it can only return: void/UpdateCount", method)
            }
        }

        val mapperName = method.resultMapperName()
        val mappings = method.parseMappingData()
        val resultSetMapper = mappings.getMapping(VertxTypes.rowSetMapper)
        val rowMapper = mappings.getMapping(VertxTypes.rowMapper)
        if (returnType.isFlow()) {
            val flowParam = returnType.arguments[0]
            val flowReturnTypeName = flowParam.toTypeName().copy(false)
            val mapperType = VertxTypes.rowMapper.parameterizedBy(flowReturnTypeName)
            if (rowMapper != null) {
                return Mapper(rowMapper, mapperType, mapperName)
            }
            return Mapper(mapperType, mapperName)
        }
        val mapperType = VertxTypes.rowSetMapper.parameterizedBy(returnTypeName)
        if (resultSetMapper != null) {
            return Mapper(resultSetMapper, mapperType, mapperName)
        }

        if (rowMapper != null) {
            return if (returnType.isList()) {
                Mapper(rowMapper, mapperType, mapperName) {
                    CodeBlock.of("%T.listRowSetMapper(%L)", VertxTypes.rowSetMapper, it)
                }
            } else {
                Mapper(rowMapper, mapperType, mapperName) {
                    CodeBlock.of("%T.singleRowSetMapper(%L)", VertxTypes.rowSetMapper, it)
                }
            }
        }

        return Mapper(mapperType, mapperName)
    }

    private fun enrichWithExecutor(repositoryElement: KSClassDeclaration, builder: TypeSpec.Builder, constructorBuilder: FunSpec.Builder) {
        builder.addProperty("_vertxConnectionFactory", VertxTypes.connectionFactory, KModifier.PRIVATE)
        builder.addSuperinterface(VertxTypes.repository)
        builder.addFunction(
            FunSpec.builder("getVertxConnectionFactory")
                .addModifiers(KModifier.OVERRIDE)
                .returns(VertxTypes.connectionFactory)
                .addStatement("return this._vertxConnectionFactory")
                .build()
        )
        val executorTag = repositoryElement.parseExecutorTag()
        if (executorTag != null) {
            constructorBuilder.addParameter(
                ParameterSpec.builder("_vertxConnectionFactory", VertxTypes.connectionFactory).addAnnotation(
                    AnnotationSpec.builder(CommonClassNames.tag).addMember("value = %L", executorTag).build()
                ).build()
            )
        } else {
            constructorBuilder.addParameter("_vertxConnectionFactory", VertxTypes.connectionFactory)
        }
        constructorBuilder.addStatement("this._vertxConnectionFactory = _vertxConnectionFactory")
    }
}
